Fixes #28148 - Add react context provider (#7139)
theforeman/foreman: webpack/stories/docs/ForemanContext.md

# ForemanContext

ForemanContext is an implementation of the new `React Context` API.
Global metadata (such as i18n, theme, foreman's settings, etc...) can be shared over all react nodes without any connection to redux nor API request.

### How to apply it

> React Router's pages contains `ForemanContext` out of the box,

Reviewer: If I'm understanding the meaning right, I think this is more clear:
```suggestion
Pages that use React Router include `ForemanContext` out of the box,
```

while a stand-alone component needs to add it in `mount_react_component`:
```ruby
  <%= mount_react_component('Component', '#selector', {
      some_data: data,
      }.to_json, { context: true }) %>

```

### How to read a value 

> Please use a constant from `/ReactApp/Context/Constants`

Reviewer: I _select which value to read_ by looking at this file, and then I use either Hooks or ForemanConsumer to consume it, right?  I think this could be more clear:
```suggestion
Available values are stored in `/ReactApp/Context/Constants`.  Read the value using one of the following methods:
```

Reviewer: Update:  Took me a minute but I think I understand better now? maybe?
* `/ReactApp/Context/Constants` is for standardizing KEY names
* ... which are SET in the `app_metadata` function in `/application_helper`, along with their values 
* So, this file should only contain names of _keys_ that are found in `/application_helper#app_metadata`.

Am I getting it now?

Reviewer: If the above is correct, I'm not sold on using constants at all for these key names.  It makes it too confusing to figure out where to set the values.  At the least, I would want to see a comment in that file clarifying that it's just for storing key names.

Updated suggestion (if we still use the constants): 
```suggestion
Keys and values are set in the `app_metadata` method of `/application_helper.rb`.  The `app_metadata` object is passed down and becomes ForemanContext.

* __Key names__ are standardized in `/ReactApp/Context/Constants` and should reflect the same keys available in `app_metadata`
* Read values using one of the following methods:
```

Author: Thanks, I used it in `How to add a value to the context` section.
I removed the constants and switched it to customs hooks

#### With Hooks

```js
// ....
import { getContext } from '../../ReactApp/Context/ForemanContext';
import { SETTINGS } from '../../ReactApp/Context/Constants';

const { perPage } = getContext(SETTINGS);
```

> #### with ForemannConsumer

Reviewer: ```suggestion
#### with ForemanConsumer
```

```js
import { ForemanConsumer } from '../../ReactApp/Context/ForemanContext';

const ThemedButton = () => (
  <ForemanConsumer>
    {({ theme }) => (
      <Button primaryColor={{ theme.primaryColor }}>
        I'm button using context!
      </Button>
    )}
  </ForemanConsumer>
);
```
### How to add a value to the context

> add a `key-value` pair to `app_metadata` function in `/application_helper` 

Reviewer: Is it also okay to add a constant to `/ReactApp/Context/Constants`?  How do these two interact?

Update: if I'm understanding right, when you add a key/value to `app_metadata` you should also add a constant in the other file.

Author: I changed it by hooks functions:
 for example, `useForemanReact()` is a custom hook that returns the version value, like selectors but for context.

Global metadata (such as version, pagination, theme, foreman's settings, etc...) can be shared over all react nodes
without any redux integration nor API request.

### How to apply it

Pages that use React Router include `ForemanContext` out of the box,
while a stand-alone component needs to add it in `mount_react_component`:
```ruby
  <%= mount_react_component('Component', '#selector', {
      some_data: data,
      }.to_json, { context: true }) %>

```

### How to read a value 

#### With Hooks
Like selectors, you can consume context values by custom hooks:

```js
// ....
import { useForemanContext } from '../../ReactApp/Context/ForemanContext';
import { SETTINGS, VERSION } from '../../ReactApp/Context/Constants';

const { perPage } = useForemanSettings();
const foremanVersion = useForemanVersion();
```

#### with ForemanConsumer
```js
import { ForemanConsumer } from '../../ReactApp/Context/ForemanContext';

const ThemedButton = () => (
  <ForemanConsumer>
    {({ docUrl }) => (
      <Button href={docUrl}>
        I'm button using context!
      </Button>
    )}
  </ForemanConsumer>
);
```
### How to add a value to the context

> Keys and values are set in the `app_metadata` method of `/application_helper.rb`.  The `app_metadata` object is passed down and becomes ForemanContext. 

Reviewer: ```suggestion
Keys and values are set in the `app_metadata` method of `/application_helper.rb`.  The `app_metadata` object is passed down and becomes ForemanContext. 

If you add a new value to `app_metadata`, it will also be useful to add a new custom hook in `webpack/assets/javascripts/react_app/ReactApp/Context/ForemanContext.js` so that the value can be easily consumed on the front end.
```

> # ForemanContext

Reviewer 2: can you use the new storybook Text component?
or convert it into an `mdx` file?
